README.md
![Screenshot](https://github.com/esmaeil-ahmadipour/esmaeil-ahmadipour/blob/main/upload/packages/flutter_glass/banner.png?raw=true "Flutter Glass Banner")
# Flutter Glass plugin

A flutter package that adds the possibility of making glass containers to your app.

![Screenshot](https://github.com/esmaeil-ahmadipour/esmaeil-ahmadipour/blob/main/upload/packages/flutter_glass/preview.gif?raw=true "Flutter Glass Demo")

### Installation

To use this plugin, add flutter_glass in your `pubspec.yaml`

```
dependencies:
  flutter_glass: ^0.0.8
```

Or install automatically using this command

```
$ flutter pub add flutter_glass
```

### Super simple to use

```dart
import 'package:flutter_glass/flutter_glass.dart';

...
    GlassContainer(
        children: [
            Padding(
                padding: const EdgeInsets.all(48.0),
                child: Text("Sample Text"),
            ),
        ],
    ),
...
```

### And bonus widget 🎉

```dart
import 'package:flutter_glass/flutter_glass.dart';

...
    GradiantContainer(
      child:  // Something ...It can be a glass container ! ,
    ),
...

```

### License

This SDK is available under the MIT license.


CHANGELOG.md
## 0.0.1
* initial release

## 0.0.2
* add gradiant widget

## 0.0.3
* improving example

## 0.0.4
* add background image for glass container

## 0.0.5
* flexible glass container size

## 0.0.6
* set opacity for glass container

## 0.0.7
* update borderRadius parameter with new object (GlassBorderSide)

## 0.0.8
* width of glass container removed from required parameters.

test/flutter_glass_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_glass/flutter_glass.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.
import 'package:flutter/material.dart';
import 'package:flutter_glass_example/main.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/lib/home.dart
import 'package:flutter/material.dart';
import 'package:flutter_glass_example/model/model.dart';
import 'package:flutter_glass_example/widgets/card_content.dart';
import 'package:flutter_glass/flutter_glass.dart';
import 'package:flutter_glass_example/widgets/list_view_body_widget.dart';

class Homepage extends StatefulWidget {
  const Homepage({Key? key}) : super(key: key);

  @override
  _HomepageState createState() => _HomepageState();
}

class _HomepageState extends State<Homepage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: SafeArea(
        child: SizedBox(
          child: GradiantContainer(  // <=== utility widget for set gradiant
              child: ListViewBody(
            children: [
              for (var i = 10; i > 0; --i)
                GlassContainer(
                  height: 200,
                  width: 360,
                  children: [
                    CardContent(
                      cardModel: CardModel(
                        cNumber: "0000 1241 3698 7845",
                        name: "User $i",
                        cvv2: "$i" "45",
                        date: "08/01",
                      ),
                    ),
                  ],
                ),
            ],
          )),
        ),
      ),
    );
  }
}


lib/flutter_glass.dart
import 'dart:ui';
import 'package:flutter/material.dart';

/// [GlassContainer] this class is main widget of this package and handle and manage all ui elements.
class GlassContainer extends StatelessWidget {
  const GlassContainer({
    Key? key,
    this.width,
    this.height,
    required this.children,
    this.borderWidth = 1.5,
    this.mainColor = Colors.white,
    this.shadowColor = Colors.black,
    this.innerPadding = const EdgeInsets.all(16.0),
    this.outerPadding = const EdgeInsets.all(8.0),
    this.blurRadius = 16.0,
    this.spreadRadius = 16.0,
    this.sigmaX = 10.0,
    this.sigmaY = 10.0,
    this.decorationImage,
    this.shadowOpacity = 0.1,
    this.mainColorOpacity = 0.2,
    this.borderRadius = const GlassBorderSide(bottomSide: 16.0, topSide: 16.0),
  }) : super(key: key);

  /// [width] , this property used for width of glass container .
  final double? width;

  /// [height] , this property used for height of glass container .
  final double? height;

  /// [borderRadius] , this property used for rounded corners in glass container .
  final GlassBorderSide borderRadius;

  /// [borderWidth] , this property set size for border around of glass container .
  final double borderWidth;

  /// [mainColor] , this property used for background color of glass container & border of this widget  (this color used in widget , by opacity 0.2) .
  final Color mainColor;

  /// [shadowColor] , this property set color of shadow in boxShadow (this color used in widget , by opacity 0.1) .
  final Color shadowColor;

  /// [innerPadding] , this property set padding to inside of  glass container .
  final EdgeInsetsGeometry innerPadding;

  /// [outerPadding] , this property set padding to outside of  glass container .
  final EdgeInsetsGeometry outerPadding;

  /// [children] , this property is REQUIRED & hold content in glass container .
  final List<Widget> children;

  /// [blurRadius] , this property used for rounded corners in boxShadow .
  final double blurRadius;

  /// [spreadRadius] , this property used for shadow impact width in boxShadow .
  final double spreadRadius;

  /// [decorationImage] , this property set background image for glass container .
  final DecorationImage? decorationImage;

  /// [sigmaX] : Users can set amount of blurring in X directions for increasing the blurring of the background .
  final double sigmaX;

  /// [sigmaY]  : Users can set amount of blurring in Y directions for increasing the blurring of the background .
  final double sigmaY;

  /// [shadowOpacity]  : this property used when users need set shadow Opacity  for outer of glass container.
  final double shadowOpacity;

  /// [mainColorOpacity]  : this property used when users need set Opacity on mainColor of glass container.
  final double mainColorOpacity;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: outerPadding,
      child: Container(
        decoration: BoxDecoration(
          boxShadow: [
            BoxShadow(
              blurRadius: blurRadius,
              spreadRadius: spreadRadius,
              color: shadowColor.withOpacity(shadowOpacity),
            )
          ],
          image: decorationImage,
          borderRadius: BorderRadius.only(
            bottomLeft: Radius.circular(borderRadius.bottomSide),
            bottomRight: Radius.circular(borderRadius.bottomSide),
            topLeft: Radius.circular(borderRadius.topSide),
            topRight: Radius.circular(borderRadius.topSide),
          ),
        ),
        child: ClipRRect(
          borderRadius: BorderRadius.only(
            bottomLeft: Radius.circular(borderRadius.bottomSide),
            bottomRight: Radius.circular(borderRadius.bottomSide),
            topLeft: Radius.circular(borderRadius.topSide),
            topRight: Radius.circular(borderRadius.topSide),
          ),
          child: BackdropFilter(
              filter: ImageFilter.blur(
                sigmaX: sigmaX,
                sigmaY: sigmaY,
              ),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Flexible(
                    child: Container(
                        width: width,
                        height: height,
                        decoration: BoxDecoration(
                          color: mainColor.withOpacity(mainColorOpacity),
                          borderRadius: BorderRadius.only(
                            bottomLeft:
                                Radius.circular(borderRadius.bottomSide),
                            bottomRight:
                                Radius.circular(borderRadius.bottomSide),
                            topLeft: Radius.circular(borderRadius.topSide),
                            topRight: Radius.circular(borderRadius.topSide),
                          ),
                          border: Border.all(
                            width: borderWidth,
                            color: mainColor.withOpacity(mainColorOpacity),
                          ),
                        ),
                        child: Padding(
                          padding: innerPadding,
                          child: Column(
                              mainAxisSize: MainAxisSize.min,
                              children: children),
                        )),
                  ),
                ],
              )),
        ),
      ),
    );
  }
}

/// [GradiantContainer] this class used  for gradiant background of glass container.
class GradiantContainer extends StatelessWidget {
  const GradiantContainer({
    Key? key,
    required this.child,
    this.colors = const [
      Colors.deepPurple,
      Colors.pink,
      Colors.red,
      Colors.orange,
    ],
    this.begin = Alignment.topCenter,
    this.end = Alignment.bottomCenter,
  })  : assert(
            !(colors.length == 0),

            /// use colors parameter in optional but send empty color list is forbidden .
            "error on colors parameter, this parameter isEmpty."
            "use colors parameter in optional but send empty color list is forbidden ."),
        super(key: key);

  /// [child] , this property is REQUIRED & used for foreground widget on the gradiant widget .
  final Widget child;

  /// [begin] , this property used for start gradiant location .
  final AlignmentGeometry begin;

  /// [end] , this property used for end gradiant location .
  final AlignmentGeometry end;

  /// [colors] , this property list of colors used in gradiant .
  final List<Color> colors;

  @override
  Widget build(BuildContext context) {
    /// [stops] contain height of colors used in gradiant widget and
    /// we set equal height size for all used colors .
    List<double> stops = [];
    for (var i = 0; i < colors.length; ++i) {
      stops.add((1.0 / (colors.length + 1)) * (i + 1));
    }

    return Container(
      height: MediaQuery.of(context).size.height,
      width: MediaQuery.of(context).size.width,
      decoration: BoxDecoration(
        gradient: LinearGradient(
          colors: colors,
          begin: begin,
          end: end,
          stops: stops,
        ),
      ),
      child: child,
    );
  }
}

/// [GlassBorderSide] this class used for rounded border of image and glass container.
class GlassBorderSide {
  /// [topSide]  , used for rounded border of top side of image and glass container
  final double topSide;

  /// [bottomSide]  , used for rounded border of top side of image and glass container
  final double bottomSide;

  const GlassBorderSide({this.topSide = 16.0, this.bottomSide = 16.0});
}


example/lib/widgets/card_content.dart
import 'package:flutter/material.dart';
import 'package:flutter_glass_example/model/model.dart';

class CardContent extends StatelessWidget {
  const CardContent({
    Key? key,
    required this.cardModel,
  }) : super(key: key);
  final CardModel cardModel;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text('My Bank ',
                style: TextStyle(color: Colors.white.withOpacity(0.75))),
            SizedBox(
                width: 48,
                height: 48,
                child:
                    Icon(Icons.payment, color: Colors.white.withOpacity(0.75))),
          ],
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(cardModel.name,
                style: TextStyle(color: Colors.white.withOpacity(0.75))),
            Text(
                'Date: '
                '${cardModel.date}',
                style: TextStyle(color: Colors.white.withOpacity(0.75))),
          ],
        ),
        const SizedBox(
          height: 120.0,
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(cardModel.cNumber,
                style: TextStyle(color: Colors.white.withOpacity(0.75))),
            Text(
                'CVV2: '
                '${cardModel.cvv2}',
                style: TextStyle(color: Colors.white.withOpacity(0.75))),
          ],
        ),
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_glass/flutter_glass.dart';
import 'package:flutter_glass_example/model/model.dart';
import 'package:flutter_glass_example/widgets/card_content.dart';
import 'package:flutter_glass_example/widgets/list_view_body_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Glass',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const Homepage(),
    );
  }
}

class Homepage extends StatelessWidget {
  const Homepage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: SafeArea(
        child: SizedBox(
          child: GradiantContainer(
              // in here, used utility & pretty widget (for set background gradiant).
              child: ListViewBody(
            children: [
              for (var i = 9; i > 0; --i)
                GlassContainer(
                  // from here, set glass widget.
                  // for using [decorationImage] , better is reduce blur on image , recommended set values of sigmaX & sigmaX on 0.0 .
                  sigmaX: 0,
                  sigmaY: 0,
                  decorationImage: const DecorationImage(
                      // in here, set your image for background  of glass widget.
                      // you can use AssetImage , NetworkImage or CachedNetworkImageProvider .
                      image: NetworkImage("http://via.placeholder.com/200x150"),
                      fit: BoxFit.fill,
                      // you can set manually opacity for background image .
                      opacity: 0.2),
                  children: [
                    // in here, put your widgets in glass widget.
                    CardContent(
                      cardModel: CardModel(
                        cNumber: "6${i}00 1241 3698 7845",
                        name: "User $i",
                        cvv2: "$i" "45",
                        date: "2022/0$i",
                      ),
                    ),
                  ],
                ),
            ],
          )),
        ),
      ),
    );
  }
}
